src/System.Private.CoreLib/shared/System/Text/ValueStringBuilder.AppendFormat.cs

```diff
@@ -0,0 +1,360 @@
+            }
+
+            AppendFormatHelper(null, format, new ParamsArray(args));
+        }
+
+        public void AppendFormat(IFormatProvider? provider, string format, object? arg0) => AppendFormatHelper(provider, format, new ParamsArray(arg0));
+
+        public void AppendFormat(IFormatProvider? provider, string format, object? arg0, object? arg1) => AppendFormatHelper(provider, format, new ParamsArray(arg0, arg1));
+
+        public void AppendFormat(IFormatProvider? provider, string format, object? arg0, object? arg1, object? arg2) => AppendFormatHelper(provider, format, new ParamsArray(arg0, arg1, arg2));
+
+        public void AppendFormat(IFormatProvider? provider, string format, params object?[] args)
+        {
+            if (args == null)
+            {
+                // To preserve the original exception behavior, throw an exception about format if both
+                // args and format are null. The actual null check for format is in AppendFormatHelper.
+                string paramName = (format == null) ? nameof(format) : nameof(args);
+                throw new ArgumentNullException(paramName);
+            }
+
+            AppendFormatHelper(provider, format, new ParamsArray(args));
+        }
+
+        private void AppendSpanFormattable<T>(T value) where T : ISpanFormattable
+        {
+            if (value.TryFormat(_chars.Slice(_pos), out int charsWritten, format: default, provider: null))
+            {
+                _pos += charsWritten;
+            }
+            else
+            {
+                string? s = value.ToString();
+                if (s != null)
+                {
+                    Append(s);
+                }
+            }
+        }
+
+        internal void AppendSpanFormattable<T>(T value, string? format, IFormatProvider? provider) where T : ISpanFormattable, IFormattable
+        {
+            if (value.TryFormat(_chars.Slice(_pos), out int charsWritten, format, provider))
+            {
+                _pos += charsWritten;
+            }
+            else
+            {
+                Append(value.ToString(format, provider));
+            }
+        }
+
+        // Copied from StringBuilder, can't be done via generic extension
+        // as ValueStringBuilder is a ref struct and cannot be used in a generic.
+        internal void AppendFormatHelper(IFormatProvider? provider, string format, ParamsArray args)
+        {
+            // Undocumented exclusive limits on the range for Argument Hole Index and Argument Hole Alignment.
+            const int IndexLimit = 1000000; // Note:            0 <= ArgIndex < IndexLimit
+            const int WidthLimit = 1000000; // Note:  -WidthLimit <  ArgAlign < WidthLimit
+
+            if (format == null)
+            {
+                throw new ArgumentNullException(nameof(format));
+            }
+
+            int pos = 0;
+            int len = format.Length;
+            char ch = '\x0';
```

Comment:
Nit: `'\0'`?

Reply:
Was just a copy paste from StringBuilder https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs#L1549

But makes sense to change.